invoices/invoiceregistry.go

```diff
@@ -427,25 +494,42 @@ func (i *InvoiceRegistry) NotifyExitHopHtlc(rHash lntypes.Hash,
 	circuitKey channeldb.CircuitKey, hodlChan chan<- interface{},
 	payload Payload) (*HodlEvent, error) {
 
-	i.Lock()
-	defer i.Unlock()
-
 	debugLog := func(s string) {
 		log.Debugf("Invoice(%x): %v, amt=%v, expiry=%v, circuit=%v",
 			rHash[:], s, amtPaid, expiry, circuitKey)
 	}
 
+	customRecords := payload.CustomRecords()
+
 	// Create the update context containing the relevant details of the
 	// incoming htlc.
 	updateCtx := invoiceUpdateCtx{
 		circuitKey:           circuitKey,
 		amtPaid:              amtPaid,
 		expiry:               expiry,
 		currentHeight:        currentHeight,
-		finalCltvRejectDelta: i.finalCltvRejectDelta,
-		customRecords:        payload.CustomRecords(),
+		finalCltvRejectDelta: i.cfg.FinalCltvRejectDelta,
+		customRecords:        customRecords,
+	}
+
+	// Process key send if present. Do this outside of the lock, because
+	// AddInvoice obtains its own lock. This is no problem, because the
+	// operation is idempotent.
+	if i.cfg.AcceptKeySend {
```

Comment:
Not a blocker, but would be nice to have an RPC mode similar to our `ChannelAcceptance` call that lets the caller examine the full payload set and reject/accept based on its contents. An example would be depositing to an exchange, they check they're actually expecting a deposit before they accept/pull the payment to avoid having to deal with refunds.

Reply:
Yes, @cfromknecht and I spoke about that too. But it is another project.